okhttp/src/main/java/okhttp3/internal/http2/Hpack.java

```diff
@@ -511,6 +521,37 @@ void writeInt(int value, int prefixMask, int bits) {
       out.writeByte(value);
     }
 
+    int getHeaderIndex(Header header) {
+      int dynamicHeaderIndex = Util.indexOf(dynamicTable, header);
+      if (dynamicHeaderIndex != -1) return normalizeDynamicTableIndex(dynamicHeaderIndex);
+
+      // Only search a subset of the static header table. Most entries have an empty value, so it's
+      // unnecessary to waste cycles looking at them.
+      for (int i = 1; i < 7; i++) {
+        if (equal(STATIC_HEADER_TABLE[i], header)) return i + 1;
+      }
+
+      return -1;
+    }
+
+    /** Converts a raw array index in the dynamic table array into the HPACK index address space. */
+    int normalizeDynamicTableIndex(int arrayIndex) {
+      return arrayIndex - nextHeaderIndex + STATIC_HEADER_TABLE.length;
+    }
+
+    int getHeaderNameIndex(ByteString name) {
+      Integer staticIndex = NAME_TO_FIRST_INDEX.get(name);
+      if (staticIndex != null) return staticIndex + 1;
+
+      for (int i = 0, size = dynamicTable.length; i < size; i++) {
```

Comment:
More stuff to consider, for now or for later...

Rather than searching the entire dynamic table (which has nulls) we could loop from `i = nextHeaderIndex + 1`  through the size. That avoids the null check here.

And we could do likewise in the code that computes `dynamicIndex` also.
```
          int dynamicIndex = Util.indexOf(dynamicTable, header);
```
Seems like this code was mostly written to do the job, and we haven’t really looked over it to make it optimal.

Also while I’m here, I was looking in `evictToRecoverBytes()` and I think the bounds of that loop are wrong.
```
        for (int j = dynamicTable.length - 1; j >= nextHeaderIndex && bytesToRecover > 0; j--) {
```
Since `nextHeaderIndex` should always be null, that check should either be `>` instead of `>=`, or even better – we can remove the check completely. As written there’s no way it ever matters.